Mathlib/Algebra/QuadraticAlgebra/Basic.lean

```diff
@@ -0,0 +1,389 @@
+@[simp]
+theorem omega_mul_coe_mul (n x y : R) :
+    (ω : QuadraticAlgebra R a b) * (↑n) * ⟨x, y⟩ = ⟨a * n * y, n * x + n * b * y⟩ := by
+  ext <;> simp; ring
+
+theorem omega_prop : (ω) * (ω) =
+    (b : QuadraticAlgebra R a b) * (ω : QuadraticAlgebra R a b) + (a : QuadraticAlgebra R a b) := by
+  ext <;> simp
+
+theorem omega_prop' : (ω : QuadraticAlgebra R a b) * (ω) =
+    b • (ω) + a • 1 := by
+  simp [omega_prop, ← QuadraticAlgebra.coe_mul_eq_smul]
+
+theorem mk_eq_add_smul_omega {x y : R} :
+    (⟨x, y⟩ : QuadraticAlgebra R a b) = x + y • (ω : QuadraticAlgebra R a b) := by
+  ext <;> simp
+
+variable {A : Type*} [Ring A] [Algebra R A]
+
+@[ext]
+theorem hom_ext {f g : QuadraticAlgebra R a b →ₐ[R] A}
+    (h : f (ω) = g (ω)) : f = g := by
+  ext ⟨x, y⟩
+  simp only [mk_eq_add_smul_omega, map_add, map_smul, h, add_left_inj]
+  change f (algebraMap R _ x) = g (algebraMap R _ x)
+  simp only [AlgHom.commutes]
+
+/-- The unique `AlgHom` from `QuadraticAlgebra R a b` to an `R`-algebra `A`,
+constructed by replacing `ω` with the provided root.
+Conversely, this associates to every algebra morphism `QuadraticAlgebra R a b →ₐ[R] A`
+a value of `ω` in `A`. -/
+@[simps]
+def lift : { u : A // u * u = b • u + a • 1 } ≃ (QuadraticAlgebra R a b →ₐ[R] A) where
+  toFun u :=
+    { toFun z := z.re • 1 + z.im • u
+      map_zero' := by simp
+      map_add' z w := by
+        simp only [re_add, im_add, add_smul, ← add_assoc]
+        congr 1
+        simp only [add_assoc]
+        congr 1
+        rw [add_comm]
+      map_one' := by simp
+      map_mul' z w := by
+        symm
+        calc
+          (z.re • (1 : A) + z.im • ↑u) * (w.re • 1 + w.im • ↑u) =
+            (z.re * w.re) • (1 : A) + (z.re * w.im) • u +
+              (z.im * w.re) • u + (z.im * w.im) • (u * u) := by
+              simp only [mul_add, mul_one, add_mul, one_mul, ← add_assoc, smul_mul_smul]
+              apply add_add_add_comm'
+          _ = (z.re * w.re) • (1 : A) + (z.re * w.im+ z.im * w.re) • u +
+                (z.im * w.im) • (u * u) := by
+              congr 1
+              simp only [add_assoc]
+              rw [← add_smul]
+          _ = (z.re * w.re) • 1 + (z.re * w.im + z.im * w.re) • u +
+                (z.im * w.im) • (b • u + a • 1) := by
+              simp [u.prop]
+          _ = (z.re * w.re + a * z.im * w.im) • 1 +
+                (z.re * w.im + z.im * w.re + b * z.im * w.im) • u := by
+              simp only [smul_add]
+              module
+            _ = (z * w).re • 1 + (z * w).im • u := by
+              simp
+      commutes' r := by
+        simp [← Algebra.algebraMap_eq_smul_one] }
+  invFun f := ⟨f (ω), by
+    simp [← map_mul, omega_mul_omega_eq, add_comm]
+    ⟩
+  left_inv r := by
+    simp
+  right_inv f := by
+    ext
+    simp
+
+end omega
+
+section omega'
+
+variable [Ring R]
+
+/-- The representative of the “other” root in the quadratic algebra.
+
+One has omega R a b + omega' R a b = b -/
+def ω' : QuadraticAlgebra R a b :=
+  ⟨b, -1⟩
```

Comment:
I think I would advocate for dropping this definition entirely, and instead standardize expressions with `star ω` instead.